plutus-use-cases/src/Plutus/Contracts/SealedBidAuction.hs

```diff
@@ -0,0 +1,304 @@
+                    .\/ Endpoint "payout" ()
+type SellerSchema = Endpoint "payout" ()
+
+-- | Definition of an auction
+data AuctionParams
+    = AuctionParams
+        { apOwner      :: PubKeyHash -- ^ Current owner of the asset. This is where the proceeds of the auction will be sent.
+        , apAsset      :: Value -- ^ The asset itself. This value is going to be locked by the auction script output.
+        , apEndTime    :: POSIXTime -- ^ When the time window for bidding ends.
+        , apPayoutTime :: POSIXTime -- ^ When the time window for revealing your bid ends.
+        }
+        deriving stock (Haskell.Eq, Haskell.Show, Generic)
+        deriving anyclass (ToJSON, FromJSON)
+
+PlutusTx.makeLift ''AuctionParams
+
+
+data SealedBid =
+    SealedBid
+        { sealedBid       :: BuiltinByteString
+        , sealedBidBidder :: PubKeyHash
+        }
+    deriving stock (Haskell.Eq, Haskell.Show, Generic)
+    deriving anyclass (ToJSON, FromJSON)
+
+PlutusTx.unstableMakeIsData ''SealedBid
+
+instance Eq SealedBid where
+  (SealedBid bid bidder) == (SealedBid bid' bidder') = bid == bid' && bidder == bidder'
+
+data RevealedBid =
+    RevealedBid
+        { revealedBid       :: Integer
+        , revealedBidBidder :: PubKeyHash
+        }
+    deriving stock (Haskell.Eq, Haskell.Show, Generic)
+    deriving anyclass (ToJSON, FromJSON)
+
+PlutusTx.unstableMakeIsData ''RevealedBid
+
+-- | The states of the auction
+data AuctionState
+    = Ongoing [SealedBid] -- Bids can be submitted.
+    | AwaitingPayout RevealedBid [SealedBid] -- The bidding is finished and we are awaiting payout
+    | Finished
+    deriving stock (Generic, Haskell.Show, Haskell.Eq)
+    deriving anyclass (ToJSON, FromJSON)
+
+PlutusTx.unstableMakeIsData ''AuctionState
+
+-- | Initial 'AuctionState'. In the beginning there are no bids.
+initialState :: AuctionState
+initialState = Ongoing []
+
+-- | Transition between auction states
+data AuctionInput
+    = PlaceBid SealedBid -- Register a sealed bid
+    | RevealBid RevealedBid -- Reveal a bid
+    | Payout
+    deriving stock (Generic, Haskell.Show)
+    deriving anyclass (ToJSON, FromJSON)
+
+PlutusTx.unstableMakeIsData ''AuctionInput
+
+type AuctionMachine = StateMachine AuctionState AuctionInput
+
+{-# INLINABLE showInteger #-}
+showInteger :: Integer -> BuiltinString
+showInteger k = if k < 0 then "-" `appendString` go (negate k) emptyString else go k emptyString
+  where
+    digits = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]
+    go n s
+      | n == 0 && s == "" = "0"
+      | n == 0            = s
+      | otherwise         = go (n `divide` 10) (digits !! (n `modulo` 10) `appendString` s)
+
+{-# INLINABLE hashInteger #-}
+hashInteger :: Integer -> BuiltinByteString
+hashInteger = sha2_256 . encodeUtf8 . showInteger
+
+{-# INLINABLE hashSecretInteger #-}
+hashSecretInteger :: Secret Integer -> BuiltinByteString
+hashSecretInteger = escape_sha2_256 . fmap (encodeUtf8 . showInteger)
```

Comment:
Did you measure the cost of this function? I'm wondering what kind of integers are viable to show on-chain

Reply:
Resolving this with reference to the two other places where the same thing is discussed :)